In [3]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

# Cargar el dataset y la matriz de interacciones procesada
df = pd.read_csv('../data/raw/SuperStoreOrders - SuperStoreOrders.csv')
matriz_interaccion = pd.read_csv('../data/matriz_interaccion.csv', index_col='customer_name')

# --- MODELO 1: BASELINE DE POPULARIDAD ---
baseline_popularidad = df.groupby(['product_id', 'product_name'])['quantity'].sum().reset_index()
baseline_popularidad = baseline_popularidad.sort_values(by='quantity', ascending=False)

print("--- TOP 5 RECOMENDACIONES (BASELINE DE POPULARIDAD) ---")
display(baseline_popularidad.head(5))

# --- MODELO 2: SIMILITUD ENTRE PRODUCTOS (COSENO) ---
matriz_productos = matriz_interaccion.T 
similitud_productos = cosine_similarity(matriz_productos)
df_similitud = pd.DataFrame(similitud_productos, index=matriz_productos.index, columns=matriz_productos.index)

print("\n--- EJEMPLO DE SIMILITUD DE PRODUCTOS ---")
primer_producto = df_similitud.columns[0]
print(f"Productos más similares a {primer_producto}:")
display(df_similitud[primer_producto].sort_values(ascending=False).head(6))

--- TOP 5 RECOMENDACIONES (BASELINE DE POPULARIDAD) ---


,product_id,product_name,quantity
2933,OFF-AR-10000110,"Binney & Smith Sketch Pad, Blue",97
3376,OFF-AR-10003651,"Sanford Pencil Sharpener, Easy-Erase",93
7594,OFF-ST-10004377,"Rogers File Cart, Single Width",89
3467,OFF-AR-10004519,"Boston Canvas, Fluorescent",86
3800,OFF-BI-10001808,"Cardinal Binding Machine, Clear",84



--- EJEMPLO DE SIMILITUD DE PRODUCTOS ---
Productos más similares a FUR-ADV-10000002:


FUR-ADV-10000002    1.000000
TEC-PH-10004626     0.894427
TEC-PH-10002780     0.848528
OFF-ST-10000718     0.732793
OFF-SU-10001968     0.539360
OFF-LA-10002364     0.537546
Name: FUR-ADV-10000002, dtype: float64

### Conclusiones y Trade-offs del Modelado (Sprint 1)
1. **Baseline de Popularidad:** Excelente para resolver el problema de *cold start* en usuarios nuevos (recomienda lo que todos compran), pero carece de personalización individual.
2. **Similitud de Coseno (Filtrado Basado en Contenido/Productos):** Captura relaciones directas entre artículos comprados conjuntamente, pero está limitado por la alta dimensionalidad y la escasez de interacciones conjuntas en catálogos grandes (*long tail*).
- **Decisión para el siguiente Sprint:** Utilizar el Baseline como sistema de respaldo y evolucionar la similitud hacia un recomendador basado en vecinos cercanos (*KNN*) o filtrado colaborativo avanzado.

otebook 03: Experimentación de Modelos (Modelado)

Propósito principal: Diseñar, ejecutar y evaluar las primeras estrategias algorítmicas de recomendación comercial basadas en la matriz de interacciones.

Baseline de Popularidad: Agrega el volumen histórico total de ventas por artículo para generar un "Top 5" de productos más vendidos, diseñado específicamente para resolver el problema del cold start (ofreciendo opciones masivas a usuarios nuevos sin historial previo).

Similitud por Coseno: Implementa un modelo matemático que calcula la cercanía entre vectores de productos dentro de la matriz, permitiendo identificar qué artículos comparten patrones de compra similares para sugerir alternativas afines.

Análisis de Trade-offs: Documenta las ventajas y limitaciones de cada enfoque técnico, destacando que el Baseline aporta estabilidad ante la falta de datos, mientras que el modelo de similitud captura relaciones directas pero enfrenta restricciones debido a la alta esparsidad del catálogo.

#Resumen Ejecutivo para Notebook 03 (Experimentación y Modelado)

Logro principal: Diseño, ejecución y evaluación de las dos primeras estrategias algorítmicas de recomendación comercial sobre la base de datos procesada.

Baseline de Popularidad: Modelo de respaldo basado en el volumen global de ventas que genera un "Top 5" de artículos más solicitados (como libretas y sacapuntas), diseñado específicamente para solucionar el problema de arranque en frío (cold start) ante usuarios nuevos sin historial de navegación.

Modelo de Similitud de Coseno: Enfoque matemático basado en el cálculo de vectores de cercanía dentro de la matriz de interacciones. Permite puntuar la afinidad entre productos (arrojando coeficientes de similitud cercanos a 0.89 u 0.84) para sugerir alternativas complementarias bajo el principio de que los clientes con hábitos similares compran artículos afines.

Trade-offs y Conclusiones Técnicas: Se documentó que el Baseline aporta estabilidad comercial ante la ausencia de datos individuales, mientras que el modelo de similitud captura relaciones directas entre ítems a pesar de las limitaciones impuestas por la alta esparsidad del catálogo.